lighthouse-core/gather/driver/target-manager.js

```diff
@@ -0,0 +1,125 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/**
+  * @fileoverview This class tracks multiple targets (the page itself and its OOPIFs) and allows consumers to
+  * listen for protocol events before each target is resumed.
+  */
+
+const log = require('lighthouse-logger');
+
+/** @typedef {{target: LH.Crdp.Target.TargetInfo, session: LH.Gatherer.FRProtocolSession}} TargetWithSession */
+
+class TargetManager {
+  /** @param {LH.Gatherer.FRProtocolSession} session */
+  constructor(session) {
+    this._enabled = false;
+    this._session = session;
+    /** @type {Array<(targetWithSession: TargetWithSession) => Promise<void>|void>} */
+    this._listeners = [];
+
+    this._onSessionAttached = this._onSessionAttached.bind(this);
+
+    /** @type {Map<string, TargetWithSession>} */
+    this._targets = new Map();
+
+    /** @param {LH.Crdp.Page.FrameNavigatedEvent} event */
+    this._onFrameNavigated = async event => {
+      // Child frames are handled in `_onSessionAttached`.
+      if (event.frame.parentId) return;
+
+      // It's not entirely clear when this is necessary, but when the page switches processes on
+      // navigating from about:blank to the `requestedUrl`, resetting `setAutoAttach` has been
+      // necessary in the past.
+      await this._session.sendCommand('Target.setAutoAttach', {
+        autoAttach: true,
+        flatten: true,
+        waitForDebuggerOnStart: true,
+      });
+    };
+  }
+
+  /**
+    * @param {LH.Gatherer.FRProtocolSession} session
+    */
+  async _onSessionAttached(session) {
+    try {
+      const target = await session.sendCommand('Target.getTargetInfo').catch(() => null);
+      const targetType = target && target.targetInfo && target.targetInfo.type;
+      const hasValidTargetType = targetType === 'page' || targetType === 'iframe';
+      if (!target || !hasValidTargetType) return;
+
+      const targetId = target.targetInfo.targetId;
+      session.setTargetInfo(target.targetInfo);
```

Comment:
could `session` take care of this itself rather than requiring an external class to query and set the targetInfo?

Reply:
it could, although there is no async initialization lifecycle on sessions today. having an async `init` that could fail to be called felt similarly brittle while also violating the lightweight connection invariant we have currently. passing the targetInfo in at the constructor was the closest I'm certainly up for, but that's still an external class querying and setting the targetInfo if that's where your concern is.

Reply:
> it could, although there is no async initialization lifecycle on sessions today

my point was that if this is the only use of subsession initialization and it's async to any observer of target-manager, might as well encapsulate `session` state in `session`, but sounds like that might be premature and the shape of things is still developing.